lib/ex_geo.ex

```diff
@@ -0,0 +1,33 @@
+defmodule ExGeo do
+  alias ExGeo.{
+    Store,
+    Result
+  }
+
+  defmodule InvalidIpError do
+    defexception message: "invalid ip address"
+  end
+
+  @type ipv4 :: {integer, integer, integer, integer}
+  @type ipv6 :: {integer, integer, integer, integer, integer, integer, integer, integer}
+  @type ip_address :: ipv4 | ipv6
+
+  defguardp is_ip_address(ip) when is_tuple(ip) and (tuple_size(ip) == 4 or tuple_size(ip) == 8)
+
+  @doc """
+  Looks up geolocation information for a binary or tuple ip
+  """
+  @spec lookup(binary | ip_address) :: Store.t
+  def lookup(ip) when is_binary(ip) do
+    String.to_charlist(ip)
+    |> :inet.parse_address()
+    |> case do
+      {:ok, ip} -> lookup(ip)
+      _ -> raise InvalidIpError, message: "Invalid ip: #{ip}"
```

Comment:
lookup should always return and ok, error tuple and the bang variant should raise. API is a bit muddled at the moment.